### DISCUSSION TITLE
Not able To Draw on Fabric.js with two pencil brushes concurrently

### DISCUSSION THREAD

**Discussion by slatemates** [fabricjs/fabric.js#8273] [General]
We are trying to build a collaborative whiteboard using fabric.js. We send mouse events (mouse up/move/down ) as well as the stroke width/color to the remote user and try to simulate the same on the remote user. For example if the user is has selected a pencil brush on his canvas , we do the same on the remote user canvas . Similarly all mouse moments are also propogated and the remote user is able to see the drawing pixel by pixel .It all works well as long the remote user is not drawing simultaneously on his own canvas. However when the remote user is also free drawing on his own canvas , the brushes (the one created for the remote user and his own freedrawing brush ) seem to mix up (colors/width etc. become equal for both brushes ) on the remote user .Do we need to use multiple canvases each for remote user or will it possible to use two pencil brushes(user himself and the remote user brush ) concurrently .

**Comment by ShaMan123** [comment]
This is exactly why discussion are there

**Comment by slatemates** [comment]
So basically based on my current understanding , when we write concurrently draw with two separate pencil brushes instances  ( both of them have a common canvas context  , and therefore they cannot have a different stroke width or color and hence the mix up ) . Please correct me if I am wrong . Is layering of canvas and then having two different instances of canvas on top of one another the only solution . Please suggest

**Comment by ShaMan123** [comment]
You are correct.
You can workaround as you mentioned.
You can look into animating a Path object instead.
Perhaps this might work as well:
https://github.com/fabricjs/fabric.js/blob/938033a869aaf899e0f11fd4b2a038abda12e43a/src/brushes/pencil_brush.class.ts#L58

**Reply by slatemates** [reply]
Thanks for the clarity regarding the common canvas  context  . As suggested I have tried looking at the code but wasnt able to gather much . I can see that the function gets called onMouseMove event in pencil brush and  the context seems to be cleared if needsFullRender returns true.Could you please guide how can that help when we try to use two pencil brushes simultaneously . Thanks  a lot .

**Reply by slatemates** [reply]
Also you suggested regarding animating a path . Could you please elaborate. That will be really helpful. Thanks again
